# Trabalho Final - Parte 3 - Planejamento

## Condições gerais

### Hardware
- Hardware posicionado dentro do datacenter do TJRS
    - Temperatura controlada em x°C (**verificar**)

### Ferramentas de controle remoto
- Configurado V-PRO para controle remoto: Reboot, BIOS access, GRUB Boot OS selection
- Ubuntu configurado com acesso remoto via protocolo RDP

### Decisões de projeto
1. Manter Turbo Boost ON
- Evidências experimentas monstram que a CPU fica em 100% de uso com a frequência de operação em sua frequência máxima
- Se o Turbo Boost estiver OFF, o máximo de frequência atingido pela CPU é a sua frequência base

2. Temperatura da CPU
- A fim de minimizar os efeitos da temperatura na frequência máxima de operação, visto que o Turbo Boost está habilitado, o desktop foi instalado dentro da Sala Cofre do Data Center do TJ-RS
- A temperatura nesta sala fica na faixa de X°C, o que auxilia a manter uma temperatura total menor na CPU do computador
- Além disso, a cada execução foi utilizado um tempo de espera de X minutos a fim de diminuir a temperatura da CPU e evitar que houvesse impacto na frequência de operação máxima.


### Pendências
- Verificar qual a frequência de operação das execuções durante cada experimento, de modo a garantir que a frequência máxima seja atingida.
 

### Explicações sobre o GIL


 

### Confirmação de versão do Python em uso

In [1]:
import sys
print(sys.version_info)

sys.version_info(major=3, minor=12, micro=3, releaselevel='final', serial=0)


In [2]:
import platform
print(platform.python_version())

3.12.3


## Etapa 1 - Implementação

-> Apenas aplicação de multiplicação de matrizes

1. Multiprocessing - Python 3.12

2. Dask

3. Threading - Python 3.14


### Numpy
- Oversubscription de Threads (NumPy vs. Multiprocessing)
- O Problema: São criados 8 processos e definida a afinidade de cada processo para um núcleo de CPU específico. No entanto, o NumPy (np.dot) por padrão é **multithreaded**. Ele tentará usar várias threads (muitas vezes, todos os núcleos disponíveis) para cada multiplicação de matriz.
- Resultado: Para os 8 processos (do Pool), cada um tentará disparar, digamos, 16 ou 24 threads (do NumPy). Isso é chamado de oversubscription. Os processos e threads competirão caoticamente pelos recursos da CPU, e as medições de tempo serão "poluídas" e inconsistentes.
- A Solução: Forçar o NumPy a usar apenas uma thread dentro de cada processo do pool. Isso garante que o controle de afinidade seja o único responsável pelo paralelismo.
- Como Corrigir: Adicionar as linhas abaixo no início do script (antes de import numpy):

```python
import os
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"] = "1"
```

- Isso diz às bibliotecas de álgebra linear (OpenBLAS, MKL) que o NumPy usa por baixo dos panos para executarem em modo single-thread. Os resultados de benchmark serão drasticamente diferentes, mas agora refletirão corretamente o desempenho do paralelismo gerenciado com multiprocessing.


### Eficiência de memória
- Em sistemas Linux, **multiprocessingv usa o modo **fork** para criar processos. Processos "filhos" herdam o espaço de memória do "pai" (modo copy-on-write).
- Com matrizes globais evita-se cópias desnecessárias.
- As matrizes globais são criadas e inicializadas antes da criação do Pool de processos.
- Para os workers (processos) apenas os índices que eles precisam processar são informados, ao invés das matrizes em si.


## Etapa 2 - Reprodutibilidade

1. GUIX / NIX

2. 

## Etapa 3 - Experimentação

-> Verificar tempos de execução com e sem turbo boost ativo ?

1. Definição de cenários
- conjunto de configurações) para executar cada implementação
- driver, governor, tipos CPUs

2. Criação do projeto experimental
- Tabela com ids de cada execução
- Informações das execuções:
    - id execução: int
    - implementação: dash, mulprocessing, non_gil
    - número da execução: int
    - id das cpus p-core: []
    - id das cpus e-core: []
    - governor: performance
    - turbo boost: enabled / disabled
    - hyper-threading: enabled / disabled 
    - driver: ?
- Tabela de experimentos embaralhada (scrambled)
- Execuções utilizando como input a tabela de experimentos embaralhada

3. Registro das execuções
- Definição de quais informações serão salvas
- Resultados salvos em arquivo com nome único: implementação_execução_data_hora

## Etapa 4 - Criação de plots

1. 

## Etapa 5 - Escrita artigo

1. Verificar modelo de artigo Latex

2. Ferramenta para escrita de latex local, estilo Overleaf

3. Imagens / plots linkadas com os experimentos ?

4. Partes do texto
- Abstract
- Introdução
- Metodologia